### PR TITLE
Allow for async presets by introducing `load` method in Neutrino class.

### PR DESCRIPTION
**Pull request by jaridmargolin**
PR to start implementation discussion around #112

### CODE REVIEW

**Comment by jaridmargolin** [comment]
Assuming we can agree on a direction, I will implement corresponding tests.

**Comment by eliperelman** [comment]
What about using something like [Denque](https://www.npmjs.com/package/denque) which makes `.use` put Promise-returning functions on a queue, which we then completely unwind? Maybe something like:

```js
use(preset, options = {}) {
  this.queue.push(() => Promise.resolve(preset(this, options)));
}

load() {
  let current;
  let previous;

  while (!this.queue.isEmpty()) {
    current = this.queue.shift();

    if (!previous) {
      previous = current();
      continue;
    }

    previous = previous.then(current);
  }

  return previous;
}
```

```js
// usage
const api = new Neutrino();

api.use(alpha);
api.use(beta);
api.use(() => /* */);

api
  .load()
  .then(() => /* */);
```

**Comment by tj** [comment]
Bluebird is a pretty big dep, I'd rather see something like https://github.com/daliwali/promise-compose personally, perf won't matter for a tiny set of promises.

**Comment by eliperelman** [comment]
@tj oooo, nice find! I just took a look at the implementation, and it's very similar to what I was prototyping locally:

```js
constructor() {
  this.queue = [];
}

use(preset, options = {}) {
  this.queue.push(() => Promise.resolve().then(() => preset(this, options)));
}

load() {
  let loading = Promise.resolve();

  if (!this.queue.length) {
    return loading;
  }

  this.queue.forEach(preset => loading = loading.then(preset));

  return loading;
}
```

**Comment by jaridmargolin** [comment]
👍 `use` / `'load` seems like a fine solution. Seems like you are moving forward with this @eliperelman? Or would you like me to continue and modify the work in this PR?

**Comment by eliperelman** [comment]
@jaridmargolin feel free to run with the comments we put here.

**Comment by eliperelman** [comment]
I was merely just seeing if the POC was viable.

File: packages/neutrino/bin/neutrino [hunk 1]
```diff
@@ -88,17 +88,9 @@ function run(command, presets) {
...
-  }
-
-  if (args.inspect) {
-    return inspect(api.getWebpackOptions());
-  }
-
-  api[command](args)
-    .then(() => process.exit(0))
+  api.load()
+    .then(() => api.config.merge(config || {}))
```
Comment (eliperelman) [inline packages/neutrino/bin/neutrino]: This should probably be `use`d prior to load for consistency:

```js
api.use(() => api.config.merge(config));
```

Also, `config` is already defaulted to an empty object on line 62.

File: packages/neutrino/bin/neutrino [hunk 2]
```diff
@@ -88,17 +88,9 @@ function run(command, presets) {
...
+    .then(() => args.inspect ?  inspect(api.getWebpackOptions()) : api[command](args))
```
Comment (eliperelman) [inline packages/neutrino/bin/neutrino]: Let's break this up into two sections:

```js
.then(() => args.inspect && inspect(api.getWebpackOptions()))
.then(() => api[command](args))
```

File: packages/neutrino/src/neutrino.js [hunk 3]
```diff
@@ -21,10 +21,17 @@ class Neutrino extends EventEmitter {
 
     this.config = new Config();
     this.options = merge(options, { root, source, output, tests, node_modules, entry });
+    this.useQueue = []
```
Comment (eliperelman) [inline packages/neutrino/src/neutrino.js]: Missing semicolon.

File: packages/neutrino/src/neutrino.js [hunk 4]
```diff
@@ -21,10 +21,17 @@ class Neutrino extends EventEmitter {
...
   }
 
   use(preset, options = {}) {
-    preset(this, options);
+    this.useQueue.push(() => preset(this, options));
+  }
+
+  load() {
+    return this.useQueue.reduce((result, task) => {
```
Comment (eliperelman) [inline packages/neutrino/src/neutrino.js]: I like the use of `reduce`, as well as the initial `Promise.resolve` to kick it off. Makes `use` leaner. :+1: 

To pass linting, change to:

```js
load() {
  return this.useQueue.reduce((result, task) => result.then(task), Promise.resolve());
}
```

File: packages/neutrino/bin/neutrino (hunk 2 above)
Comment (jaridmargolin) [inline packages/neutrino/bin/neutrino]: I was under the impression inspect was supposed to stop execution of `api[command](args)`:

```
-  if (args.inspect) {
-    return inspect(api.getWebpackOptions());
-  }
```

File: packages/neutrino/bin/neutrino (hunk 1 above)
Comment (jaridmargolin) [inline packages/neutrino/bin/neutrino]: Cool stuffs 👍

File: packages/neutrino/bin/neutrino (hunk 2 above)
Comment (eliperelman) [inline packages/neutrino/bin/neutrino]: It's mostly a facade, `inspect` does a `process.exit`, see line 14.

Comment (jaridmargolin) [inline packages/neutrino/bin/neutrino]: Hmmmm... I'm not 100% in love with breaking them up. For someone else consuming the code, I don't think it is immediately apparent that inspect results in an exit.

Keeping them together in some sort of if/else or ternary shows that execution is one or the other. Ultimately your call. Just my 2 cents.

File: packages/neutrino/src/neutrino.js (hunk 4 above)
Comment (jaridmargolin) [inline packages/neutrino/src/neutrino.js]: Good catch. No need for assigning the result variable as I am already returning the result in reduce 👍

**Comment by eliperelman** [comment]
I had a thought, and I don't think using a loop is going to work:

```js
api.use(alpha);
api.use(neutrino => new Promise(resolve => {
  neutrino.use(beta);
  resolve();
});
api.use(gamma);
```

Since the `beta` use isn't added to the stack until it is executed, it won't exist in the queue until all items have been pushed to the queue, which only works for top-level `use`s. Probably going to need to take a recursion approach. I'll think about it more and see if there's anything else I'm missing.

**Comment by jaridmargolin** [comment]
Didn't realize this was a use case that needed to be satisfied? What would the expected behavior be in your opinion?

alpha
inline
gamma
beta

or

alpha
inline
gamma
beta

---

This to me is where the queue gets messy. If we went back to the original idea of just having use return a Promise, I think we would avoid a world of hurt.

**Comment by eliperelman** [comment]
Having use return a Promise seems like a non-starter. That means we would have to do this in middleware:

```js
Promise
  .resolve()
  .then(() => neutrino.use(alpha))
  .then(() => neutrino.use(beta))
  .then(() => neutrino.use(gamma));
```

Using Ramda's `pipeP` wouldn't help much here either.

**Comment by jaridmargolin** [comment]
I think the code is significantly more predictable when `use` returns a promise. Is your issue solely with verbosity?

async/await would significantly clean it up (uncertain of your targeted node versions).

```
await neutrino.use(alpha);
await neutrino.use(beta);
await neutrino.use(gamma);
```

**Comment by eliperelman** [comment]
Support is LTS right now (Node.js v6.9+).

**Comment by jaridmargolin** [comment]
I think you either support async and return a promise, or you stay sync. Execution order of nested `use` calls with a queue is going to be ugly and unpredictable.

For the average user consuming neutrino, they probably won't need be dealing directly with the `use` method. The added burden of an async `use` method is shifted to the preset & middleware authors. I personally don't find it bothersome, but I see you do.

Obviously I have a use case where I would prefer it to be async, so I would lean in that direction... but I understand if this isn't a concern for everyone.

**Comment by jaridmargolin** [comment]
Quickly going back to your example:

```
api.use(alpha);
api.use(neutrino => new Promise(resolve => {
  neutrino.use(beta);
  // inline config
  resolve();
});
api.use(gamma);
```

In the queue solution, this could be interpreted several ways, with the authors intent probably being the 3rd (which I'm not even sure would be possible).

alpha
inline
gamma
beta

or

alpha
inline
beta
gamma

or

alpha
beta
inline
gamma

**Comment by eliperelman** [comment]
@jaridmargolin I can't think of a reasonable solution to this either. I think we can move forward with this if you feel it meets your use case without imposing any additional burden on existing middleware or sync presets. Thoughts?

Also there are a few failing tests in Travis. :smiley:

File: packages/neutrino/bin/neutrino [hunk 5]
```diff
@@ -88,17 +88,10 @@ function run(command, presets) {
...
+  api.use(() => api.config.merge(config));
 
-  if (args.inspect) {
-    return inspect(api.getWebpackOptions());
-  }
-
-  api[command](args)
-    .then(() => process.exit(0))
+  api.load()
+    .then(() => args.inspect ?  inspect(api.getWebpackOptions()) : api[command](args))
```
Comment (eliperelman) [inline packages/neutrino/bin/neutrino]: Remove extra space after `?`.

**Comment by jaridmargolin** [comment]
> I think we can move forward with this

Not sure I understand. The queue implementation has way to many edge cases to introduce (IMO). The only reliable solution I see to offer async support is to alter the `use` command to return a resolved Promise (which you at one point called a "non-starter"). This will absolutely have implications for preset/middleware authors as mentioned above:

```
neutrino.use(alpha)
  .then(() => neutrino.use(beta))
  .then(() => neutrino.use(gamma));
  ...
```

Looking at the current presets in neutrino-dev, I could see how this could be undesired. Async/await would make the transition significantly easier, but it seems that would not be possible due to currently supported versions of node.

Now while the syntax would be temporarily clunky, if you plan to support this in the future, it seems like now would be the time to make the change. The farther you kick the can down the road, the more presets/middleware will require breaking changes in the future.

Or the decision can be made that `use` will always be synchronous, and any async work would need to be completed prior to running neutrino.

---

How do you want to proceed?

```
a. Update `use` to return a Promise
b. Close PR and require presets/middleware to execute sync
```

**Comment by eliperelman** [comment]
Sorry, I meant move forward with this PR as it is.

**Comment by eliperelman** [comment]
@helfi92 @tj @SpencerCDixon any opinions for moving Neutrino v5 support to Node 7 w/ async await support?

**Comment by eliperelman** [comment]
We could also add a pipe method which could just be passed presets to load in series:

```js
neutrino.pipe(alpha, beta, gamma);
```

**Comment by jaridmargolin** [comment]
@eliperelman - The current solution, and pipe, would both suffer from the same problem with nested calls to `use`, correct?

Because the ordering of execution effects override precedence, I think it would make more sense for the execution order to be simple/expected than to support async use cases.

**Comment by eliperelman** [comment]
Yes, you are correct. I think for now let's punt on async presets and possibly revisit when Node.js v8 goes LTS. Thanks for putting so much though into this! (And if you think of another idea, let's talk about it.)

**Comment by SpencerCDixon** [comment]
>  any opinions for moving Neutrino v5 support to Node 7 w/ async await support?

+1 for waiting for 8 and LTS

**Pull request closed without merging**